docs/hpc/advanced/debugging_mpi_programs/gdb_mpi_debug.md

```diff
@@ -0,0 +1,228 @@
+
+4. **Backtracing**: When an error occurs, GDB can provide a backtrace of function calls leading up to the error, helping developers pinpoint the root cause of the issue.
+
+5. **Memory Inspection**: GDB allows developers to inspect memory contents, including stack frames, heap allocations, and register values, aiding in the detection of memory-related errors such as buffer overflows or memory leaks.
+
+6. **Conditional Breakpoints**: Developers can set breakpoints that trigger only when certain conditions are met, allowing for more precise debugging in complex scenarios.
+
+7. **Core File Analysis**: GDB can analyze core dump files generated when a program crashes, providing insight into the state of the program at the time of the crash.
+
+Again, we do not aim to teach the basics of GDB in this page and we assume that the user is familiar with the basic usage. You can keep the [GDB cheat sheet](https://darkdust.net/files/GDB%20Cheat%20Sheet.pdf) open which will help you to learn its usage faster and it will also be useful in this tutorial.
+
+## Debugging MPI Programs with GDB
+
+The process of debugging MPI programs with GDB is slightly more complex than debugging serial programs due to the parallel nature of MPI applications. When debugging an MPI program, you are dealing with multiple processes running concurrently, each with its own memory space and execution context. GDB provides features to debug MPI programs, but it requires additional steps and considerations to effectively debug parallel applications.
+
+Consider that you have an MPI program as shown below. The sample program given below is simple MPI program that performs addition and multiplication on each rank and prints the result. The first argument to the multiplication functions is the rank of the process and the second argument is the size of the communicator (this is actually the bug  which is there in the code below and our task is to use gdb on such programs to find the issue). We have create this  simple  program using functions so that you can step inside while debugging. Let us assume that the file is named `mpi_debug.cpp`.
+ 
+```c++
+#include <iostream>
+#include <mpi.h>
+
+// Function for multiplication
+int multiplication(int a, int b, int rank, const char* hostname) {
+    std::cout << "Rank " << rank << " on host " << hostname << " performing multiplication with arguments: " << a << " and " << b << std::endl;
+    int result = a * b;
+    std::cout << "Result of multiplication: " << result << std::endl;
+    return result;
+}
+
+int main(int argc, char** argv) {
+    MPI_Init(&argc, &argv);
+    
+    int rank, size;
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+
```

Comment:
```suggestion
    MPI_Get_processor_name (hostname, &hostnamelen);
```

Reply:
I will use the correct code and push it again so that we can merge it.